test/apps/iodemo/io_demo.cc

```diff
@@ -177,9 +180,9 @@ class MemoryPool : public ObjectPool<BufferType, use_offcache> {
 public:
     MemoryPool(size_t buffer_size, const std::string &name,
                size_t offcache = 0) :
-        ObjectPool<BufferType, use_offcache>::ObjectPool(buffer_size, name,
-                                                         offcache)
+        ObjectPool<BufferType, use_offcache>::ObjectPool(buffer_size, name)
     {
+        this->init(offcache);
```

Comment:
this-> not needed

Reply:
without this->, compiler reported an error which says init is not within the scope and suggests use this->init instead. I guess the reason might be that init() is defined in base class, so it should be referenced by adding this-> in derived class's constructor.

Reply:
ok, maybe it's because base class is templated. can you pls post the compiler error?

Reply:
CXX      io_demo-io_demo.o
io_demo.cc: In constructor ‘MemoryPool<BufferType, use_offcache>::MemoryPool(size_t, const string&, size_t)’:
io_demo.cc:185:22: error: there are no arguments to ‘init’ that depend on a template parameter, so a declaration of ‘init’ must be available [-fpermissive]
         init(offcache);
                      ^
io_demo.cc:185:22: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
io_demo.cc: In instantiation of ‘MemoryPool<BufferType, use_offcache>::MemoryPool(size_t, const string&, size_t) [with BufferType = P2pDemoCommon::IoMessage; bool use_offcache = false; size_t = long unsigned int; std::string = std::basic_string<char>]’:
io_demo.cc:677:80:   required from here
io_demo.cc:185:22: error: ‘init’ was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
io_demo.cc:185:22: note: declarations in dependent base ‘ObjectPool<P2pDemoCommon::IoMessage, false>’ are not found by unqualified lookup
io_demo.cc:185:22: note: use ‘this->init’ instead
io_demo.cc: In instantiation of ‘MemoryPool<BufferType, use_offcache>::MemoryPool(size_t, const string&, size_t) [with BufferType = P2pDemoCommon::SendCompleteCallback; bool use_offcache = false; size_t = long unsigned int; std::string = std::basic_string<char>]’:
io_demo.cc:677:80:   required from here
io_demo.cc:185:22: error: ‘init’ was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
io_demo.cc:185:22: note: declarations in dependent base ‘ObjectPool<P2pDemoCommon::SendCompleteCallback, false>’ are not found by unqualified lookup
io_demo.cc:185:22: note: use ‘this->init’ instead
io_demo.cc: In instantiation of ‘MemoryPool<BufferType, use_offcache>::MemoryPool(size_t, const string&, size_t) [with BufferType = P2pDemoCommon::BufferIov; bool use_offcache = false; size_t = long unsigned int; std::string = std::basic_string<char>]’:
io_demo.cc:677:80:   required from here
io_demo.cc:185:22: error: ‘init’ was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
io_demo.cc:185:22: note: declarations in dependent base ‘ObjectPool<P2pDemoCommon::BufferIov, false>’ are not found by unqualified lookup
io_demo.cc:185:22: note: use ‘this->init’ instead
io_demo.cc: In instantiation of ‘MemoryPool<BufferType, use_offcache>::MemoryPool(size_t, const string&, size_t) [with BufferType = DemoServer::IoWriteResponseCallback; bool use_offcache = false; size_t = long unsigned int; std::string = std::basic_string<char>]’:
io_demo.cc:900:64:   required from here
io_demo.cc:185:22: error: ‘init’ was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
io_demo.cc:185:22: note: declarations in dependent base ‘ObjectPool<DemoServer::IoWriteResponseCallback, false>’ are not found by unqualified lookup
io_demo.cc:185:22: note: use ‘this->init’ instead
io_demo.cc: In instantiation of ‘MemoryPool<BufferType, use_offcache>::MemoryPool(size_t, const string&, size_t) [with BufferType = DemoClient::IoReadResponseCallback; bool use_offcache = false; size_t = long unsigned int; std::string = std::basic_string<char>]’:
io_demo.cc:1302:64:   required from here
io_demo.cc:185:22: error: ‘init’ was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
io_demo.cc:185:22: note: declarations in dependent base ‘ObjectPool<DemoClient::IoReadResponseCallback, false>’ are not found by unqualified lookup
io_demo.cc:185:22: note: use ‘this->init’ instead
make[3]: *** [io_demo-io_demo.o] Error 1

Reply:
if this->init is OK, I will commit another patch to only address the default value removal comment.